source/Octopus.Client.Tests/Integration/OctopusClient/BuildServerHeadersBaseTests.cs

```diff
@@ -0,0 +1,61 @@
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Nancy;
+using NUnit.Framework;
+using Octopus.Client.Model;
+
+namespace Octopus.Client.Tests.Integration.OctopusClient
+{
+    public abstract class BuildServerHeadersBaseTests : HttpIntegrationTestBase
+    {
+        protected BuildServerHeadersBaseTests()
+            : base(UrlPathPrefixBehaviour.UseClassNameAsUrlPathPrefix)
+        {
+            Get(TestRootPath, p =>
+            {
+                var automationContext = Request.Headers.UserAgent.Split(' ')[1];
+
+                return Response.AsJson(new TestDto { AutomationContext = automationContext })
+                    .WithStatusCode(HttpStatusCode.OK);
+            });
+        }
+
+        protected abstract string EnvironmentVariableName { get; }
+        protected abstract string EnvironmentVariableValue { get; }
+        internal abstract BuildEnvironment ExpectedBuildEnvironment { get; }
+
+        protected override void SetupEnvironmentVariables()
+        {
+            OctopusCustomHeaders.GetEnvironmentVariable = variableName => variableName == EnvironmentVariableName ? EnvironmentVariableValue : null;
```

Comment:
this is a little bit of a brain bender

Reply:
Yes. Yes it is. I had to do this, because it relies on Environment variables that our build servers will always set to be the ones for TeamCity, because that's what they are :) So the live code uses the real `Environment.GetEnvironmentVariable` as that function delegate and the tests can override to mock the environment variables